### PR TITLE
Preserve rp filter

### PR DESCRIPTION
**Pull request by clareconstantine**
We now save the application dashboard's regional partner filter in session storage and in redux.

- As part of that development, `RegionalPartnerFilter` now refers to the `{value: 1, label: "Regional Partner Name"}` object. (Previously, it only referred to the `value` property)

- By default, the `RegionalPartnerDropdown` updates the application dashboard's regional partner filter, but if an `onChange` method is passed to the `RegionalPartnerDropdown`, that will be run on change instead.

### CODE REVIEW

File: apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx [hunk 1]
```diff
@@ -51,7 +51,8 @@ const paths = {
 
 export default class ApplicationDashboard extends React.Component {
   static propTypes = {
-    regionalPartnerName: PropTypes.string,
+    regionalPartnerLabel: PropTypes.string,
+    regionalPartnerValue: PropTypes.number,
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx]: Any reason we want to pass these as 2 separate props rather than a single object (or hash in Ruby)?

We already have a [RegionalPartnerFilterPropType](https://github.com/code-dot-org/code-dot-org/pull/21752/files#diff-80b18c36c138da5da34afbb3c41b5313R92)

Also why do we need to supply this here? Are we conflating the regional partner filter (only relevant to admins) and the actual regional partner info (for a regional partner)? Isn't that handled by the underlying APIs? Do we need to store or pass this info for regional partners? I'm unclear why this is needed.

File: apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx [hunk 2]
```diff
@@ -62,8 +63,9 @@ export default class ApplicationDashboard extends React.Component {
   };
 
   componentWillMount() {
-    if (this.props.regionalPartnerName) {
-      store.dispatch(setRegionalPartnerName(this.props.regionalPartnerName));
+    if (this.props.regionalPartnerLabel && this.props.regionalPartnerValue) {
+      const regionalPartnerFilter = {value: this.props.regionalPartnerValue, label: this.props.regionalPartnerLabel};
+      store.dispatch(setRegionalPartnerFilter(regionalPartnerFilter));
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx]: If we make this a single prop as suggested above, then this can condense to

```es6
if (this.props.regionalPartnerFilter) {
  store.dispatch(setRegionalPartnerFilter(regionalPartnerFilter));
}
```

File: apps/src/code-studio/pd/application_dashboard/cohort_calculator.jsx [hunk 3]
```diff
@@ -11,7 +11,7 @@ const styles = {
 export default class CohortCalculator extends React.Component {
   static propTypes = {
     role: PropTypes.string.isRequired,
-    regionalPartnerFilter: PropTypes.oneOfType([
+    regionalPartnerValue: PropTypes.oneOfType([
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/cohort_calculator.jsx]: nit: it's a little confusing that we sometimes call it a filter and other times don't. I think this should be called `regionalPartnerFilterValue`.

Alternatively it could be the entire `regionalPartnerFilter` object named as such, and then the value can be referenced by `this.props.regionalPartnerFilter.value`

File: apps/src/code-studio/pd/application_dashboard/cohort_view.jsx [hunk 4]
```diff
@@ -107,20 +103,21 @@ class CohortView extends React.Component {
...
-          <CohortCalculator
-            role={this.props.route.role}
-            regionalPartnerFilter={this.state.regionalPartnerFilter}
-            accepted={accepted}
-            registered={registered}
-          />
+          {this.state.applications &&
+            <CohortCalculator
+              role={this.props.route.role}
+              regionalPartnerValue={this.props.regionalPartnerFilter.value}
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/cohort_view.jsx]: `regionalPartnerFilterValue`

File: apps/src/code-studio/pd/application_dashboard/cohort_calculator.jsx (hunk 3 above)
Comment (aoby) [reply apps/src/code-studio/pd/application_dashboard/cohort_calculator.jsx]: Note also we could also export `RegionalPartnerFilterValuePropType` to reuse this `oneOfType`.

File: apps/src/code-studio/pd/application_dashboard/constants.js [hunk 5]
```diff
@@ -75,15 +76,23 @@ exports.ApplicationFinalStatuses = [
...
-  {value: allPartnersFilter, label: allPartnersLabel}
+  {value: unmatchedValue, label: unmatchedLabel},
+  {value: allPartnersValue, label: allPartnersLabel}
 ];
+
+exports.RegionalPartnerFilterPropType = PropTypes.shape({
+  value: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.number
+  ]),
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/constants.js]: When `value` is a string, it's not just any string, it's `all` or `none` as defined above right?
So this would be better / stricter as:

```es6
value: PropTypes.oneOfType([
  PropTypes.number, // regional partner id
  PropTypes.oneOf([allPartnersValue, unmatchedValue])
]),
```

Now I also realize, those constant names are inconsistent ^. How about renaming the latter to `unmatchedPartnerValue` ?

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx [hunk 6]
```diff
@@ -181,17 +183,17 @@ export class DetailViewContents extends React.Component {
...
+    this.setState({ regional_partner_name, regional_partner_value});
   };
 
   handleSaveClick = () => {
     let stateValues = [
       'status',
       'locked',
       'notes',
-      'regional_partner_filter',
+      'regional_partner_value',
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx:198]: Why do we need this in state here? Isn't this handled by Redux now?

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx [hunk 7]
```diff
@@ -240,8 +242,8 @@ export class DetailViewContents extends React.Component {
       return (
         <RegionalPartnerDropdown
           onChange={this.handleRegionalPartnerChange}
-          regionalPartnerFilter={this.state.regional_partner_filter}
-          additionalOptions={[{label: UnmatchedLabel, value: UnmatchedFilter}]}
+          regionalPartnerFilter={{value: this.state.regional_partner_value, label: this.state.regional_partner_name}}
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx:247]: Why do we need this prop? Isn't this handled by Redux now?

File: apps/src/code-studio/pd/application_dashboard/quick_view.jsx [hunk 8]
```diff
@@ -85,13 +89,13 @@ export class QuickView extends React.Component {
     }
   }
 
-  load(regionalPartnerFilter = this.state.regionalPartnerFilter) {
+  load(regionalPartnerValue = this.props.regionalPartnerFilter.value) {
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/quick_view.jsx]: nit: like above, `regionalPartnerFilterValue`

File: apps/src/code-studio/pd/application_dashboard/regional_partner_dropdown.jsx [hunk 9]
```diff
@@ -17,16 +19,16 @@ const styles = {
...
 export class RegionalPartnerDropdown extends React.Component {
   static propTypes = {
+    // Parents can pass in an onChange function; otherwise the dropdown will
+    // update the application dashboard's regional partner filter on change
     onChange: PropTypes.func,
-    regionalPartnerFilter: PropTypes.oneOfType([
-      PropTypes.string,
-      PropTypes.number
-    ]),
+    updateRegionalPartnerFilter: PropTypes.func.isRequired,
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/regional_partner_dropdown.jsx]: Why is `onChange` needed now?

Or we can keep `onChange`, supply its value from Redux's `setRegionalPartnerFilter`, and remove `updateRegionalPartnerFilter`. In any case, I don't think we don't need 2 update props.

File: apps/src/code-studio/pd/application_dashboard/regional_partner_dropdown.jsx [hunk 10]
```diff
@@ -36,14 +38,22 @@ export class RegionalPartnerDropdown extends React.Component {
     }
   }
 
+  onChange = (selected) => {
+    if (this.props.onChange) {
+      this.props.onChange(selected);
+    } else {
+      this.props.updateRegionalPartnerFilter(selected);
+    }
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/regional_partner_dropdown.jsx]: See above. This seems unnecessary. Also note, the latter prop is required but we won't call it here when the optional `onChange` is provided.

File: apps/src/code-studio/pd/application_dashboard/regional_partner_dropdown.jsx [hunk 11]
```diff
@@ -36,14 +38,22 @@ export class RegionalPartnerDropdown extends React.Component {
...
   render() {
     return (
       <FormGroup>
         <ControlLabel>Select a regional partner</ControlLabel>
         <Select
           clearable={false}
-          value={this.props.regionalPartnerFilter}
-          onChange={this.props.onChange}
+          value={this.props.regionalPartnerFilter.value}
+          onChange={this.onChange}
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/regional_partner_dropdown.jsx]: Presuming we don't need both props (see above) this can simply pass the handle update prop (whatever we name it) directly to Select's `onChange`.

File: apps/test/unit/code-studio/pd/application_dashboard/cohort_calculatorTest.js [hunk 12]
```diff
@@ -7,11 +7,11 @@ import sinon from 'sinon';
 describe("Cohort Calculator", () => {
   describe("Initially", () => {
     let cohortCalculator;
-    const regionalPartnerFilter = "all";
+    const regionalPartnerValue = "all";
```
Comment (aoby) [inline apps/test/unit/code-studio/pd/application_dashboard/cohort_calculatorTest.js]: note we can import this from `constants` (AllPartnersValue)

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx (hunk 6 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx:198]: Since this is the instance of the dropdown used to assign an application to a regional partner (editing the app in detail view), it doesn't use Redux because it shouldn't update the filter for the application dashboard.

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx (hunk 7 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx:247]: We want to use the value stored in the component's state because it might not be the same as what is in the application dashboard filter - this instance of the dropdown is used to assign an application to a regional partner (editing the app in detail view).

File: apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx (hunk 1 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx]: For a regional partner viewing the dashboard, they will see an appropriately filtered list of applications (provided by the api) but we still need to know the regional partner's name to display. I think we could just be passing in the name though, or getting the name off the `current_user`

File: apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx [hunk 13]
```diff
@@ -62,8 +63,8 @@ export default class ApplicationDashboard extends React.Component {
   };
 
   componentWillMount() {
-    if (this.props.regionalPartnerName) {
-      store.dispatch(setRegionalPartnerName(this.props.regionalPartnerName));
+    if (this.props.regionalPartnerFilter.value && this.props.regionalPartnerFilter.label) {
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx]: What if only one or the other is provided?

File: apps/src/code-studio/pd/application_dashboard/cohort_view.jsx [hunk 14]
```diff
@@ -107,20 +103,21 @@ class CohortView extends React.Component {
       }
       return (
         <div>
-          <CohortCalculator
-            role={this.props.route.role}
-            regionalPartnerFilter={this.state.regionalPartnerFilter}
-            accepted={accepted}
-            registered={registered}
-          />
+          {this.state.applications &&
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/cohort_view.jsx:106]: Why is this needed? The spinner will be displayed while loading. After load, in the case where we have no applications (for a given partner), with this added line we'll display nothing. Wouldn't we rather display empty results so it's clear we have no applications rather than a glitch?

File: apps/src/code-studio/pd/application_dashboard/cohort_view.jsx [hunk 15]
```diff
@@ -107,20 +103,21 @@ class CohortView extends React.Component {
...
+              regionalPartnerFilterValue={this.props.regionalPartnerFilter.value}
+              accepted={accepted}
+              registered={registered}
+            />
+          }
           {this.props.isWorkshopAdmin &&
             <RegionalPartnerDropdown
-              onChange={this.handleRegionalPartnerChange}
-              regionalPartnerFilter={this.state.regionalPartnerFilter}
+              regionalPartnerFilter={this.props.regionalPartnerFilter}
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/cohort_view.jsx]: Why override this property? This is the Redux-connected version of `RegionalPartnerDropdown`, which will get the value from redux. I realize we're overriding it with the same value, so it won't cause any issues, but it's confusing and unnecessary. Unless I'm missing something?

File: apps/src/code-studio/pd/application_dashboard/constants.js [hunk 16]
```diff
@@ -75,15 +76,25 @@ exports.ApplicationFinalStatuses = [
...
+const allPartnersValue = "all";
+const unmatchedPartnerLabel = "No Partner/Unmatched";
+const unmatchedPartnerValue = "none";
 exports.AllPartnersLabel = allPartnersLabel;
-exports.AllPartnersFilter = allPartnersFilter;
-exports.UnmatchedLabel = unmatchedLabel;
-exports.UnmatchedFilter = unmatchedFilter;
+exports.AllPartnersValue = allPartnersValue;
+exports.UnmatchedPartnerLabel = unmatchedPartnerLabel;
+exports.UnmatchedPartnerValue = unmatchedPartnerValue;
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/constants.js:85]: nit: I know this has come up before, but this seems unwieldy. Unless we have a compelling reason to duplicate all these constants, how about defining and exporting at once like this:
```es6
export const AllPartnersLabel = "All Regional Partners' Applications";
export const AllPartnersValue = "all";
// etc
```

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx [hunk 17]
```diff
@@ -181,17 +185,17 @@ export class DetailViewContents extends React.Component {
   };
 
   handleRegionalPartnerChange = (selected) => {
-    const regional_partner_filter = selected ? selected.value : UnmatchedFilter;
-    const regional_partner_name = selected ? selected.label : UnmatchedLabel;
-    this.setState({ regional_partner_name, regional_partner_filter});
+    const regional_partner_id = selected ? selected.value : UnmatchedPartnerValue;
+    const regional_partner_name = selected ? selected.label : UnmatchedPartnerLabel;
+    this.setState({ regional_partner_name, regional_partner_id});
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx]: This part doesn't appear to be new, so I assume it works, but what if `all` or `none` come through as selected?

File: apps/src/code-studio/pd/application_dashboard/quick_view.jsx [hunk 18]
```diff
@@ -85,13 +89,13 @@ export class QuickView extends React.Component {
     }
   }
 
-  load(regionalPartnerFilter = this.state.regionalPartnerFilter) {
+  load(regionalPartnerFilterValue = this.props.regionalPartnerFilter.value) {
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/quick_view.jsx]: It appears `load` has this filter passed in both places it's called. Why do we need the default value?

File: dashboard/app/controllers/api/v1/pd/applications_controller.rb [hunk 19]
```diff
@@ -53,8 +53,8 @@ def quick_view
       role = params[:role].to_sym
       applications = get_applications_by_role(role)
 
-      unless params[:regional_partner_filter].blank? || params[:regional_partner_filter] == 'all'
-        applications = applications.where(regional_partner_id: params[:regional_partner_filter] == 'none' ? nil : params[:regional_partner_filter])
+      unless params[:regional_partner_value].blank? || params[:regional_partner_value] == 'all'
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/applications_controller.rb]: Can we use the constant, `REGIONAL_PARTNERS_ALL`?

File: dashboard/app/controllers/api/v1/pd/applications_controller.rb [hunk 20]
```diff
@@ -53,8 +53,8 @@ def quick_view
...
+        applications = applications.where(regional_partner_id: params[:regional_partner_value] == 'none' ? nil : params[:regional_partner_value])
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/applications_controller.rb]: same as above, can we use `REGIONAL_PARTNERS_NONE`?

File: dashboard/app/controllers/api/v1/pd/applications_controller.rb [hunk 21]
```diff
@@ -78,14 +78,14 @@ def quick_view
...
     def cohort_view
       role = params[:role]
-      regional_partner_filter = params[:regional_partner_filter]
+      regional_partner_value = params[:regional_partner_value]
       applications = get_applications_by_role(role.to_sym).where(status: ['accepted', 'withdrawn'])
 
-      unless regional_partner_filter.nil? || regional_partner_filter == 'all'
-        applications = applications.where(regional_partner_id: regional_partner_filter == 'none' ? nil : regional_partner_filter)
+      unless regional_partner_value.nil? || regional_partner_value == 'all'
+        applications = applications.where(regional_partner_id: regional_partner_value == 'none' ? nil : regional_partner_value)
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/applications_controller.rb]: and here

File: apps/src/code-studio/pd/application_dashboard/constants.js (hunk 16 above)
Comment (aoby) [reply apps/src/code-studio/pd/application_dashboard/constants.js:85]: Also we should consider naming them like `ALL_PARTNERS_LABEL` to conform to our styleguide, though we're pretty inconsistent on this point.

Comment (aoby) [reply apps/src/code-studio/pd/application_dashboard/constants.js:85]: After some discussion, and consulting our styleguide, we should label them uppercase, e.g.
```es6
export const ALL_PARTNERS_VALUE = "all";
```

File: apps/src/code-studio/pd/application_dashboard/quick_view.jsx (hunk 18 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/quick_view.jsx]: We don't!

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx (hunk 17 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx]: `all` isn't an option here, but if `none` is selected then it will be used as the `regional_partner_id`. I was leaning towards calling it `regional_partner_id` throughout this file because that is what the property is in the application data. I can change it to use `regional_partner_value` though to be more clear that it can be something other than an id.

File: apps/src/code-studio/pd/application_dashboard/cohort_view.jsx (hunk 15 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/cohort_view.jsx]: oh you're right - I just missed this one. thanks!

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx (hunk 17 above)
Comment (aoby) [reply apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx]: I like that: either call it `regional_partner_id` and have it always be an id or `null`, or allow `"none"` and call it `_value`

File: apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx (hunk 13 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/application_dashboard.jsx]: then we won't have enough info to use the filter so we shouldn't set it, and it should use the default values in the reducer (Unmatched).

File: apps/src/code-studio/pd/application_dashboard/cohort_view.jsx (hunk 14 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/cohort_view.jsx:106]: This line prevents the cohort calculator from showing up empty. The table of applications will still appear empty. I added this to match the behavior we have on list view.

File: apps/src/code-studio/pd/application_dashboard/constants.js (hunk 16 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/constants.js:85]: i'll follow up with a PR fixing these and other constants
